packages/braid-design-system/src/lib/components/TooltipRenderer/TooltipRenderer.tsx

```diff
@@ -56,38 +55,72 @@ export const TooltipTextDefaultsProvider = ({
 };
 
 export const TooltipContent = ({
-  inferredPlacement,
-  arrowLeftOffset,
+  placement,
+  arrowPosition,
+  arrowRef,
   children,
 }: {
-  inferredPlacement: PopoverProps['placement'];
-  arrowLeftOffset: ArrowLeftOffset;
+  placement: PopoverProps['placement'];
+  arrowPosition?: { x?: number; y?: number };
+  arrowRef?: React.RefObject<HTMLElement | null>;
   children: ReactNodeNoStrings;
-}) => (
-  <Box
-    textAlign="left"
-    boxShadow="large"
-    background="neutral"
-    borderRadius="large"
-    padding="small"
-    marginX={edgeOffset}
-    className={[styles.maxWidth, styles.translateZ0]}
-  >
-    <TooltipTextDefaultsProvider>
-      <Box className={styles.overflowWrap} zIndex={1} position="relative">
-        {children}
-      </Box>
-      <Box
-        position="fixed"
-        background="neutral"
-        className={styles.arrow[inferredPlacement!]}
-        style={assignInlineVars({
-          [styles.horizontalOffset]: `${arrowLeftOffset}px`,
-        })}
-      />
-    </TooltipTextDefaultsProvider>
-  </Box>
-);
+}) => {
+  const arrowX = arrowPosition?.x;
+  const arrowY = arrowPosition?.y;
+
+  const { space, grid } = useSpace();
+  const edgeOffsetInPx = grid * space.xsmall;
+  const arrowWidthOffset = parseFloat(styles.constants.arrowSize) * 2;
+
+  const [tooltipWidth, setTooltipWidth] = useState(0);
+  const tooltipContainerRef = useRef<HTMLElement | null>(null);
+
+  const handleTooltipContainerRef = (element: HTMLElement | null) => {
+    tooltipContainerRef.current = element;
+    if (element) {
+      const rect = element.getBoundingClientRect();
+      setTooltipWidth(rect.width);
+    }
+  };
+
+  const clampedArrowX =
+    arrowX !== undefined && tooltipWidth > 0
+      ? clamp(
+          edgeOffsetInPx,
+          arrowX,
+          tooltipWidth - edgeOffsetInPx - arrowWidthOffset,
+        )
+      : arrowX;
+
+  return (
+    <Box
+      ref={handleTooltipContainerRef}
+      textAlign="left"
+      boxShadow="large"
+      background="neutral"
+      borderRadius="large"
+      padding="small"
+      className={[styles.maxWidth, styles.translateZ0]}
+    >
+      <TooltipTextDefaultsProvider>
+        <Box className={styles.overflowWrap} zIndex={1} position="relative">
+          {children}
+        </Box>
+        <Box
+          ref={arrowRef}
+          position="fixed"
+          background="neutral"
+          className={styles.arrow[placement ?? 'top']}
+          style={assignInlineVars({
+            [styles.arrowX]:
+              clampedArrowX !== undefined ? `${clampedArrowX}px` : 'unset',
+            [styles.arrowY]: arrowY !== undefined ? `${arrowY}px` : 'unset',
```

Comment:
Rather than setting the value to `unset`, is there a reason you are setting the var at all?

```suggestion
              clampedArrowX !== undefined ? `${clampedArrowX}px` : undefined,
            [styles.arrowY]: arrowY !== undefined ? `${arrowY}px` : undefined,
```